Initial document on AssemblyLoadContext Design

gkhanna79 opened the pull request:
skip ci please

@jkotas PTAL

jkotas commented:
@agocke @davidfowl @gokarnm Could you please take a look and tell whether it has answers to everything you ever wanted to know about AssemblyLoadContext? Thanks!

jkotas commented on Documentation/design-docs/assemblyloadcontext.md:
```diff
@@ -0,0 +1,63 @@
+
+### LoadFromAssemblyName
+
+This method can be used to load an assembly into a load context different from the current one.
```
I would say:

This method can be used to load an assembly using a load context different from the load context of currently executing assembly.

gkhanna79 replied:
Fixed.

jkotas commented on Documentation/design-docs/assemblyloadcontext.md:
```diff
@@ -0,0 +1,63 @@
+
+## API Surface
+
+Most of the **LoadContex** [API surface](https://github.com/dotnet/corefx/blob/master/src/System.Runtime.Loader/ref/System.Runtime.Loader.cs) is self-explanatory. Key APIs/Properties, though, are described below:
```
It would be useful to explain the relationship between the classic APIs `Assembly.Load`, `Assembly.LoadFrom`, `Assembly.LoadFile`, `Assembly.Load(byte[])` and AssemblyLoadContexts. What is the load context used by each of these APIs.

gkhanna79 replied:
@jkotas Is this question in context of the APIs we have exposed for NS 2.0?

jkotas replied:
Yes. @rahku may be the best person to fill it in since he has done the work on it.

agocke replied:
Also, an explanation for why no `Assembly.Load(byte[])` but an `Assembly.Load(Stream)` would be useful.

I ended up wrapping an `ImmutableArray<byte>` in a `MemoryStream` but it seems a little weird to have to do that.

jkotas replied:
Stream is more flexible/efficient. `byte[]` can be wrapped in MemoryStream easily, but the other direction cannot be done efficiently.

`Load(byte[])` requires the assembly to be stored in the exact right sized byte array. If you have the assembly in something else, you have to pay the cost of allocating (potentially large) `byte[]`.

gkhanna79 replied:
> Yes. @rahku may be the best person to fill it in since he has done the work on it.

Had chatted with @rahku about this. Document is now updated to reflect this.

jkotas commented on Documentation/design-docs/assemblyloadcontext.md:
```diff
@@ -0,0 +1,63 @@
+
+### Default
+
+This property will return a reference to the *Default LoadContext*.
```
Explain where the list of the assemblies for the default load context comes from?

gkhanna79 replied:
I mentioned its composition at the start of the writeup.

jkotas commented:
LGTM otherwise. Thank you for writing this!

agocke commented on Documentation/design-docs/assemblyloadcontext.md:
```diff
@@ -0,0 +1,63 @@
+
+### Resolution Process
+If the assembly was already present in *A1's* context, either because we had successfully loaded it earlier or because we failed to load it for some reason, we return the corresponding status (and assembly reference for the success case).
+However, if *C1* was not found in *A1's* context, the *Load* method override in *A1's* context is invoked. 
```
Maybe it should be noted whether or not `mscorlib` is an exception to this? My understanding is that mscorlib can never be reloaded.

gkhanna79 replied:
Done.

agocke commented on Documentation/design-docs/assemblyloadcontext.md:
```diff
@@ -0,0 +1,63 @@
+## Constraints
+
+* Currently, custom **LoadContext** cannot be unloaded once created. This is a feature we are looking into for a future release.
+* If an attempt is made to load a R2R image from the same location in multiple load context's, then precompiled code can only be used from the first image that got loaded. The subsequent images will have their code JITted. This happens because subsequent loading binaries from the same location results in OS mapping them to the same memory as the previous one was mapped to and thus, could corrupt  ...
```
What about if the custom load context requests the load from another load context? e.g.,

```csharp
protected override Assembly Load(AssemblyName name) => Default.LoadFromAssemblyName(name);
```

agocke replied:
My observation is that doing this allows types (unification and static values) to persist if this is done, so that may be an interesting thing to call out.

gkhanna79 replied:
If LC_A attempts to resolve from LC_B, which already has the assembly loaded, then there is no second load attempted since the assembly is already found in LC_B (See the process section). This is really about reloading the assembly within a different context.

jkotas replied:
> My observation is that doing this allows types (unification and static values)

Yes, you can build a graph (typically a tree) between different AssemblyLoadContext to control what gets unified.

davidfowl commented on Documentation/design-docs/assemblyloadcontext.md:
```diff
@@ -0,0 +1,73 @@
+
+If the *Default LoadContext* fallback also did not resolve the load (or was not applicable), the *Resolving* event is invoked against *A1's* load context. This is the last oppurtunity to attempt to resolve the assembly load. If there are no subscribers for this event, or neither resolved the load, a *FileNotFoundException* is thrown.
+
+## PInvoke Resolution
```
Should we add something about resource assemblies?

gkhanna79 replied:
What specific thing about resource assemblies are you looking for @davidfowl ?

agocke commented:
One more request: could we get some info about the use of `SetProfileOptimizationRoot`?

gokarnm commented on Documentation/design-docs/assemblyloadcontext.md:
```diff
@@ -0,0 +1,73 @@
+### Basics
+If an assembly *A1* triggers the load of an assembly *C1*, the latter's load is attempted within the **LoadContext** instance of the former (which is also known as the *RequestingAssembly* or *ParentAssembly*). 
+
+Dynamically generates assemblies add a slight twist since they do not have a *ParentAssembly/RequestingAssembly* per-se. Thus, they are associated with the load context of their *Creator Assembly* and any subsequent loads (static or dynamic) will use that load context.
```
Typo _Dynamically generates assemblies_ => _Dynamically generated assemblies_

gokarnm commented on Documentation/design-docs/assemblyloadcontext.md:
```diff
@@ -0,0 +1,73 @@
+
+As part of .NET Standard 2.0 effort, certain assembly load APIs off the **AppDomain** type, which were present in Desktop .NET Framework, have been brought back. The following maps the APIs to the load context in which they will load the assembly:
+
+* Assembly.Load - loads the assembly into the *Default LoadContext*
```
I'm a bit concerned with how the `Assembly.Load` and `Assembly.LoadFrom` seem to be "custom ALC unfriendly", similar to `AssemblyLoadContext.Default`, as a developer can use it to write code that works fine under default ALC (the common case) but potentially break when the same code is run under a custom ALC.

As someone who is developing a plugin framework leveraging the ALC feature, I'm looking at how third party framework developers can be directed towards the "pit of success" where the code they write works seamlessly under a custom ALC.

Maybe having a "Best Practices" section about the writing ALC friendly code, or/and the API documentation for these three APIs mentioning the pitfall of using default context would be helpful. Something on the lines of _"This API always loads the assembly in the default Load Context. Consider using this [alternate API] if your code is not tied to the default load context and can run in any load context"_.

Some inputs for the proposed "Best Practices" section

1. When to override Load method vs. implement Resolve event handler for a custom ALC based on @gkhanna79 advice on this [thread](https://github.com/dotnet/coreclr/issues/8398#issuecomment-264235647).

1. `AssemblyLoadContext.Default` returns the reference to the default load context, use `AssemblyLoadContext.GetLoadContext(Assembly.GetExecutingAssembly())` if you want a reference to the current executing assembly's load context.

1. `Assembly.Load` and `Assembly.LoadFrom` always load the assembly into the default load context. If your code is not tied to the default load context and can be run in any load context, consider using `AssemblyLoadContext.Load` or `AssemblyLoadContext.LoadFromAssemblyPath` methods on the executing assembly's load context. You can get the executing assembly's load context using `AssemblyLoadContext.GetLoadContext(Assembly.GetExecutingAssembly())`.

davidfowl replied:
I thought Assembly.Load used the "current" load context. If it doesn't how would any of this work for plugins?

gokarnm replied:
The latest version of the document says that `Assembly.Load` loads the assembly into the context of the assembly that triggers the load, which addresses my concern and seems to be the correct behavior.
What about `Assembly.Assembly.LoadFrom` though? Does it load into the default context or the "current" context?

natemcmaster commented on Documentation/design-docs/assemblyloadcontext.md:
```diff
@@ -0,0 +1,73 @@
+If the assembly was already present in *A1's* context, either because we had successfully loaded it earlier or because we failed to load it for some reason, we return the corresponding status (and assembly reference for the success case).
+However, if *C1* was not found in *A1's* context, the *Load* method override in *A1's* context is invoked. 
+
+* For *Custom LoadContext*, this override is an oppurtunity to load an assembly **before** the fallback (see below) to *Default LoadContext* is attempted to resolve the load. 
```
Doesn't resolution first check the default load context? Observed behavior is that Load on *A1*'s context is not called if the default load context already contains an assembly identity matching *C1*. https://github.com/dotnet/coreclr/issues/6187

Also, spelling: oppurtunity => opportunity

natemcmaster commented on Documentation/design-docs/assemblyloadcontext.md:
```diff
@@ -0,0 +1,73 @@
+
+### Resolution Process
+If the assembly was already present in *A1's* context, either because we had successfully loaded it earlier or because we failed to load it for some reason, we return the corresponding status (and assembly reference for the success case).
+However, if *C1* was not found in *A1's* context, the *Load* method override in *A1's* context is invoked. 
```
Doesn't resolution first check the default load context? Observed behavior is that Load on *A1*'s context is not called if the default load context already contains an assembly identity matching *C1*. Doc bug or implementation bug? cref https://github.com/dotnet/coreclr/issues/6187

gkhanna79 replied:
Mentioned at https://github.com/dotnet/coreclr/pull/10271/files#diff-5adec7752d04e4f6a6b40f455af2ee55R29

natemcmaster replied:
> If the assembly was already present in *A1's* context, either because we had successfully loaded it earlier or because we failed to load it for some reason, we return the corresponding status (and assembly reference for the success case).

I read that sentence differently then. I understood "successfully loaded" to mean "successfully loaded _in A1's context._" 

So you're saying it means "successfully loaded _in A1's load context or the default context_", or is it "successfully loaded _in any load context_"?

gkhanna79 replied:
This sentence is in the context of A1's LC.

natemcmaster commented on Documentation/design-docs/assemblyloadcontext.md:
```diff
@@ -0,0 +1,73 @@
+
+* **System.Private.CoreLib.dll** is only loaded once, and into the **Default LoadContext**, during the .NET Core Runtime startup as it is a logical extension of the same. It cannot be loaded into **Custom LoadContext**.
+* Currently, custom **LoadContext** cannot be unloaded once created. This is a feature we are looking into for a future release.
+* If an attempt is made to load a R2R image from the same location in multiple load context's, then precompiled code can only be used from the first image that got loaded. The subsequent images will have their code JITted. This happens because subsequent loading binaries from the same location results in OS mapping them to the same memory as the previous one was mapped to and thus, could corrupt  ...
```
Can you define "R2R image" in this document?

gkhanna79 replied:
Done.

natemcmaster commented on Documentation/design-docs/assemblyloadcontext.md:
```diff
@@ -0,0 +1,73 @@
+
+## API Surface
+
+Most of the **AssemblyLoadContext** [API surface](https://github.com/dotnet/corefx/blob/master/src/System.Runtime.Loader/ref/System.Runtime.Loader.cs) is self-explanatory. Key APIs/Properties, though, are described below:
```
> Most of the AssemblyLoadContext API surface is self-explanatory.

This sentence is meaningless. I suggest removing it.

gkhanna79 replied:
This is present to explain why this document is not explaining functionality of all the APIs.

gkhanna79 commented:
Needless to say, when I write when I am half-asleep, I miss on few things. Now that I am wide-awake, I have pushed an update that I hope addresses most of the comments (for the ones not addressed, I have put up questions for the comment author).

gkhanna79 commented on Documentation/design-docs/assemblyloadcontext.md:
```diff
@@ -0,0 +1,73 @@
+## PInvoke Resolution
+
+*Custom LoadContext* can override the **AssemblyLoadContext.LoadUnmanagedDll** method to intercept PInvokes from within the **LoadContext** instance so that can be resolved from custom binaries. If not overridden, or if the resolution is not able to resolve the PInvoke, the default PInvoke mechanism will be used as fallback.
+
```
Done.

gkhanna79 commented:
> One more request: could we get some info about the use of SetProfileOptimizationRoot?

I will merge this PR and have @brianrob add context around it.

gkhanna79 commented:
LoadFrom loads into the default context.



Thanks,
Gaurav


-------- Original message --------
From: Milind Gokarn <notifications@github.com>
Date: 3/20/17 1:55 PM (GMT-08:00)
To: dotnet/coreclr <coreclr@noreply.github.com>
Cc: "Gaurav Khanna (CLR)" <Gaurav.Khanna@microsoft.com>, State change <state_change@noreply.github.com>
Subject: Re: [dotnet/coreclr] Initial document on AssemblyLoadContext Design (#10271)


@gokarnm commented on this pull request.

________________________________

In Documentation/design-docs/assemblyloadcontext.md<https://github.com/dotnet/coreclr/pull/10271#discussion_r107014051>:

> +
+Most of the **AssemblyLoadContext** [API surface](https://github.com/dotnet/corefx/blob/master/src/System.Runtime.Loader/ref/System.Runtime.Loader.cs) is self-explanatory. Key APIs/Properties, though, are described below:
+
+### Default
+
+This property will return a reference to the *Default LoadContext*.
+
+### LoadFromAssemblyName
+
+This method can be used to load an assembly into a load context different from the load context of the currently executing assembly.
+
+## Assembly Load APIs and LoadContext
+
+As part of .NET Standard 2.0 effort, certain assembly load APIs off the **AppDomain** type, which were present in Desktop .NET Framework, have been brought back. The following maps the APIs to the load context in which they will load the assembly:
+
+* Assembly.Load - loads the assembly into the *Default LoadContext*


The latest version of the document says that Assembly.Load loads the assembly into the context of the assembly that triggers the load, which addresses my concern and seems to be the correct behavior.
What about Assembly.Assembly.LoadFrom though? Does it load into the default context or the "current" context?

—
You are receiving this because you modified the open/close state.
Reply to this email directly, view it on GitHub<https://github.com/dotnet/coreclr/pull/10271#discussion_r107014051>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AJA3wcFFeoGrLL4rCOB0PaVQP6Kzf-Hwks5rnue6gaJpZM4MhK3q>.

gokarnm commented:
Hi @gkhanna79, I'm curious about why `Assembly.Load` and `Assembly.LoadFrom' have different behavior in terms of the load context they load into, could you provide more insight? If `Assembly.LoadFrom' uses default context, how would you use it with plugins, it could cause potential issues when used from code running in a custom ALC?

gkhanna79 commented:
@gokarnm Assembly.Load* APIs have been (re)introduced with respect to the Desktop .NET Framework.

- Assembly.Load, in .NET Core, works exactly how it works on Desktop.
- Assembly.LoadFrom, on Desktop, worked differently based upon certain heuristics that gave inconsistent/non-deterministic results. Sometimes, based upon rules, it would load in Default Context and at other times, it would load in its own _LoadFrom_ context. When we had enabled this for our internal infrastructure, when bringing up Windows Phone, we realized that simply loading in the default context didn't actually break anyone and worked as expected. Thus, for .NET Core, we chose to adopt the same semantic, making it more predictable.

Plugins written to use these APIs should expect to behave the same way as they did on the Desktop (afterall, these APIs are introduced from Desktop Compat standpoint as well). If someone wishes to load in a particular context, then they should use ALC APIs.

gokarnm commented:
@gkhanna79 , thanks for the details!

In Desktop .NET Framework, AppDomain was commonly used for implementing plugin frameworks and developers would have relatively less knowledge of load contexts. I would appreciate if the API documentation for `Assembly.LoadFrom` had a note, something like - 
_"This API always loads the assembly into the default load context. To load your assemblies in a particular context, use APIs on the AssemblyLoadContext type."_ 

That would partially address my concern that developers would use `Assembly.LoadFrom` as they are familiar with it from .NET Framework, but it may not be the right API in some scenarios, causing their code to break when run under a custom load context.

gkhanna79 commented:
> I would appreciate if the API documentation for Assembly.LoadFrom had a note, something like

This is called out at the very end of the section of the Assembly Load APIs.

gokarnm commented:
>This is called out at the very end of the section of the Assembly Load APIs.

I noticed that, my thinking was - developers may not even know about ALC and read this document, or the ALC API reference. If the [API reference](https://docs.microsoft.com/en-us/dotnet/core/api/system.reflection.assembly) for `Assembly.LoadFrom` had a similar note, it would be more discoverable. Maybe some of this content is planned to flow into API reference? Sorry for my ignorance.

gkhanna79 commented:
@danmosemsft Can the above be tracked as part of documentation update for APIs exposed out of NS 2.0?

danmosemsft commented:
@karelz is tracking documentation...

karelz commented:
Please file a bug - either here in CoreFX, or in [docs repo](https://github.com/dotnet/docs/issues) (BTW: @mairaw which one do you prefer?)

Ideally: Link the docs page. Suggest changes (mention text before and after).
Once we review it for correctness, we will ask docs team to make the changes.

mairaw commented:
No preference but https://github.com/dotnet/docs seems more appropriate for doc issues.

ro-jo commented:
When I search Bing or Google for AssemblyLoadContext documentation, it leads me to this page: #5463 

I read through the whole post looking for a link to some documentation, maybe I missed it.  Then I found an unqualified link to this discussion.

I haven't got time to read every word in this discussion or other discussions: where is the documentation for AssemblyLoadContext please?

Is this it, surely not?: https://docs.microsoft.com/en-gb/dotnet/api/system.runtime.loader.assemblyloadcontext?view=netcore-2.0

BrainCrumbz commented:
For future ref., this should be the "dedicated" doc page you were looking for: https://github.com/dotnet/coreclr/blob/master/Documentation/design-docs/assemblyloadcontext.md
